// @ark/cli-xhs/dist/post-xhs.js
import { execSync } from 'node:child_process';
export function buildFullBody(body, tags) {
    const tagsText = tags.map(t => `#${t}`).join(' ');
    return `${body}\n\n${tagsText}`;
}
function run(cmd) {
    return execSync(cmd, { encoding: 'utf8' });
}
function runInherit(cmd) {
    execSync(cmd, { stdio: 'inherit' });
}
function findRef(snapshot, label) {
    // label can be a plain word (e.g. "写长文") or exact pattern (e.g. 'button "发布"')
    const escaped = label.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    const regex = new RegExp(`[^\\n]*${escaped}[^\\n]*\\[ref=(e\\d+)\\]`);
    const match = regex.exec(snapshot);
    if (!match)
        throw new Error(`Could not find "${label}" in snapshot:\n${snapshot}`);
    return match[1];
}
export async function postToXhs(input, sessionPath) {
    const fullBody = buildFullBody(input.body, input.tags);
    const title = input.title.replace(/"/g, '\\"');
    const body = fullBody.replace(/"/g, '\\"');
    // 1. Restart daemon with session, then open publish page
    try {
        runInherit(`agent-browser close`);
    }
    catch { /* ignore if not running */ }
    runInherit(`agent-browser --state "${sessionPath}" open "https://creator.xiaohongshu.com/publish/publish"`);
    runInherit(`agent-browser wait --load networkidle`);
    // 2. Click 写长文
    let snapshot = run(`agent-browser snapshot -i`);
    const changwenRef = findRef(snapshot, '写长文');
    runInherit(`agent-browser click "@${changwenRef}"`);
    runInherit(`agent-browser wait --load networkidle`);
    // 3. Click 新的创作
    snapshot = run(`agent-browser snapshot -i`);
    const newRef = findRef(snapshot, '新的创作');
    runInherit(`agent-browser click "@${newRef}"`);
    runInherit(`agent-browser wait --load networkidle`);
    // 4. Fill title and body
    runInherit(`agent-browser fill "[placeholder*='标题']" "${title}"`);
    runInherit(`agent-browser fill "[contenteditable='true']" "${body}"`);
    // 5. Click 一键排版
    snapshot = run(`agent-browser snapshot -i`);
    const layoutRef = findRef(snapshot, '一键排版');
    runInherit(`agent-browser click "@${layoutRef}"`);
    runInherit(`agent-browser wait --load networkidle`);
    // 6. Click 下一步
    snapshot = run(`agent-browser snapshot -i`);
    const nextRef = findRef(snapshot, '下一步');
    runInherit(`agent-browser click "@${nextRef}"`);
    runInherit(`agent-browser wait --load networkidle`);
    // 7. Click 发布 (match exact button text, not "暂存离开" etc.)
    snapshot = run(`agent-browser snapshot -i`);
    const publishRef = findRef(snapshot, 'button "发布"');
    runInherit(`agent-browser click "@${publishRef}"`);
    runInherit(`agent-browser wait --load networkidle`);
    // 8. Get final snapshot to extract post URL
    const finalSnapshot = run(`agent-browser snapshot`);
    const urlMatch = /url:\s*(https?:\/\/[^\s]+)/i.exec(finalSnapshot);
    const postUrl = urlMatch?.[1] ?? 'https://www.xiaohongshu.com';
    return {
        postUrl,
        publishedAt: new Date().toISOString(),
    };
}
